### PR TITLE
added live build status to Telescope dashboard:

### LINKED ISSUE #2416: Add live build status details in status dashboard
In #2395 we are adding an expanded status service dashboard. One thing I want to include in it is the ability to see build status info from the autodeployment server, which is exposed at `/status`, for example:

- staging is https://dev.telescope.cdot.systems:4000/status
- production is https://telescope.cdot.systems:4000/status

@dhillonks started improving the code related to this in #2392.  Now I want to use it to go further. This returns a JS object with previous, current, and pending build details.  Any/all of these can be `null`, or a full JS object that we cache from the GitHub API.

Our dashboard should show some of this data in a visual way.  Things that might be interesting are:

- the git SHA that is/was built
- who pushed the code (i.e., `pusher`)
- who wrote the code (i.e., `sender` or maybe `author`, not sure)
- a link to the `compare` branch to see what changed 
- when the build started (`startedDate`)
- when the build stopped (`stoppedDate`)
- what the result code of the build was (e.g., 0 is "good" and non-zero is an "error")

### PR DESCRIPTION
**Pull request by mqnguyen5**
## Issue This PR Addresses

Fixes #2416 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

- Added scripts for fetching and rendering build-header
- Added html template for build-header
- Imported build-header to run on load

Screen Shot 2021-12-08 at 1 01 37 PM
Screen Shot 2021-12-08 at 1 22 59 PM
Screen Shot 2021-12-08 at 1 29 28 PM

To test the feature locally, do the following:

- Change `autodeploymentUrl` in `public/js/buildl-log/api.js`

```js
const autodeploymentUrl = (path) => `//dev.api.telescope.cdot.systems:4000/${path}`;  
```

- Allow all `connectSrc` in `src/server.js`

```js
connectSrc: ["'self'", "*"]
```

- Change `current` to `previous` inside `src/api/status/public/js/build-log/api.js`

```js
export const checkBuildStatus = async () => {
  try {
    const res = await fetch(autodeploymentUrl('status'));
    if (!res.ok) {
      throw new Error('unable to get build info');
    }
    const data = await res.json();
    if (!data.current) {
      return { building: false };
    }
    return {
      building: true,
      title: data.type,
      githubData: data.current.githubData,
      startedAt: new Date(data.current.startedDate),
      stoppedAt: new Date(data.current.stopDate),
      result: data.code,
    };
  } catch (err) {
    console.error(err);
    return { building: false };
  }
};
```

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by Andrewnt219** [comment]
Can you update the branch and add an instruction for testing this feature?

File: src/api/status/public/js/build-log/build-header.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      resMsg.innerHTML += 'Unable to get build info.';
+
+      throw new Error('Unable to get build info');
+    }
+    const { previous } = await res.json();
+    renderBuildInfo(previous);
+  } catch (err) {
+    console.error(err);
+  }
+}
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:47]: We don't need to double fetch from "/status". The `checkBuildStatus` is already called every 5 seconds. You can just pass down the `githubData` and call `renderBuildInfo` inside `checkForBuild` when there is an ongoing build.

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]

File: src/api/status/public/js/build-log/build-header.js (hunk 1 above)
Comment (Andrewnt219) [reply src/api/status/public/js/build-log/build-header.js:47]: That'll also keep the build-log and build-header in sync.

Comment (mqnguyen5) [reply src/api/status/public/js/build-log/build-header.js:47]: Previously, I tried to follow #2486 to test if the data was fetch successfully, but now I'm getting this error:

Screen Shot 2021-12-08 at 2 00 40 PM

Comment (Andrewnt219) [reply src/api/status/public/js/build-log/build-header.js:47]: Content Security Error, you are missing this change.  But this is only for testing, do not include it in the commit.
```diff
-         connectSrc: [
-          "'self'",
-          '*.fontawesome.com',
-          `${host.replace(/(^\w+:|^)\/\//, '')}:4000`,
-          '*.github.com',
-        ],
+ connectSrc: ["'self'", "*"]

```

**Comment by Andrewnt219** [comment]
@humphd Since there's no way to run the autodeployment locally, maybe it's easier to point this to the stagging or prod autodployment? 

https://github.com/Seneca-CDOT/telescope/blob/38d74beb90624d16103287e5e44db1b22f932595/src/api/status/public/js/build-log/api.js#L2

**Comment by humphd** [comment]
@Andrewnt219 great suggestion.  For reviewing this, that works, and we can change it back before we merge.  @mqnguyen5 can you point it at https://dev.telescope.cdot.systems for now?

**Comment by mqnguyen5** [comment]
Will do

File: src/api/status/public/js/build-log/build-header.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      icon.className = 'fas fa-server px-2';
+      resMsg.innerHTML = '';
+      resMsg.append(icon);
+      resMsg.innerHTML += 'Unable to get build info.';
+
+      throw new Error('Unable to get build info');
+    }
+    renderBuildInfo(data);
+  } catch (err) {
+    console.error(err);
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:41]: No need for a trycatch here.

File: src/api/status/public/js/build-log/build-header.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+const buildSenderName = document.getElementById('build-sender-name');
+const buildSenderImg = document.getElementById('build-sender-img');
+const buildGitSHA = document.getElementById('build-git-sha');
+const buildResult = document.getElementById('build-result');
+const buildStarted = document.getElementById('build-started');
+const buildDuration = document.getElementById('build-duration');
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  resMsg.toggleAttribute('hidden');
+  buildHeaderInfo.toggleAttribute('hidden');
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:13]: This makes the build header switch between loading state and info state every 5s. Like, 5s of info and 5s of loading.

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]

**Comment by Yoda-Canada** [comment]
I have changed ` current ` to ` previous ` inside src/api/status/public/js/build-log/api.js, but can't get the result which have the data. It only display loading status. How to get the previous information?
image

image

**Comment by mqnguyen5** [comment]
change inside the `if` statement as well

**Comment by manekenpix** [comment]
We really need to get better at rebasing instead of merging master into our branches

image

**Comment by Yoda-Canada** [comment]
I changed, but the result is same.
image

**Comment by mqnguyen5** [comment]
@Yoda-Canada sorry, I just updated the instruction to test this feature. You mind trying it again?

**Comment by mqnguyen5** [comment]
> We really need to get better at rebasing instead of merging master into our branches
> 
> image

Sorry about that, you mind showing me how to do it?

File: src/api/status/public/js/build-log/build-header.js (hunk 3 above)
Comment (mqnguyen5) [reply src/api/status/public/js/build-log/build-header.js:13]: You have any suggestions on how to fix it? I've tried using `setAttribute` and `removeAttribute` here but it doesn't seems to do the trick

Comment (Andrewnt219) [reply src/api/status/public/js/build-log/build-header.js:13]: Maybe the function should only render the build info, not toggling between two states (i.e. only render info state)? I don't think there's a case where toggling between those two is necessary.

File: src/api/status/public/js/build-log/check-for-build.js [hunk 4]
```diff
@@ -42,4 +43,5 @@ export default async function checkForBuild() {
       reader.read().then(processLog).catch(finish);
     }
   }
+  buildHeader(status);
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/check-for-build.js:46]: This needs to move up before the `if(build)`, so even when there is an ongoing build, the header will still be updated.

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]

**Comment by Andrewnt219** [comment]
> you mind showing me a better way to do it instead of git pull upstream master?

Use `git pull --rebase upstream master`

**Comment by Yoda-Canada** [comment]
@mqnguyen5 I followed your new steps, and it only displayed the loading page.

**Comment by Andrewnt219** [comment]
@mqnguyen5 Handlebars PR is merged, can you rebase this? The changes in build.html go into builds.hbs

**Comment by mqnguyen5** [comment]
Will do, I'll try to get it done soon.

**Comment by Andrewnt219** [comment]
> @mqnguyen5 I followed your new steps, and it only displayed the loading page. Do you use a specific address to get the following data?

@Yoda-Canada Have you changed all the `current` to `previous`?

**Comment by Andrewnt219** [comment]
@Yoda-Canada  can you show me the console on the build page? And you can reply in a new comment, editing my comment looks weird.

**Comment by Yoda-Canada** [comment]
@Andrewnt219 
image

**Comment by Andrewnt219** [comment]
There's no `autodeploymentUrl` in build-header.js, you need to pull the new changes.

**Comment by Yoda-Canada** [comment]
Ever though I have allowed  ` connectSrc: ["'self'", "*"] `, still can't fetch the data

**Comment by humphd** [comment]
When you say you added it, the error above shows `localhost:4000` but you're fetching from `http://dev.api.telescope.cdot.systems:4000/status` so it's not added.

**Comment by Kevan-Y** [comment]
> I followed the instruction and changed everything. I think the code has an issue, right?

Look okay on my end.

**Review by Kevan-Y — COMMENTED** [review]
image
Would it be better to have a bit more space on the left?

File: src/api/status/public/js/build-log/api.js [hunk 5]
```diff
@@ -15,7 +16,10 @@ export const checkBuildStatus = async () => {
     return {
       building: true,
       title: data.type,
+      githubData: data.current.githubData,
       startedAt: new Date(data.current.startedDate),
+      stoppedAt: new Date(data.current.stopDate),
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/api.js:21]: There's no `stopDate` for current, use the current date instead.

File: src/api/status/public/js/build-log/build-header.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+const buildSenderName = document.getElementById('build-sender-name');
+const buildSenderImg = document.getElementById('build-sender-img');
+const buildGitSHA = document.getElementById('build-git-sha');
+const buildResult = document.getElementById('build-result');
+const buildStarted = document.getElementById('build-started');
+const buildDuration = document.getElementById('build-duration');
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  buildHeaderTitle.innerHTML = '';
+  buildHeaderTitle.innerText = 'Build Status';
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/build-header.js:13]: overwriting `innerHtml` should be enough
```suggestion
  buildHeaderTitle.innerHTML = 'Build Status';
```

File: src/api/status/src/views/builds.hbs [hunk 7]
```diff
@@ -32,6 +32,43 @@
     </div>
   </nav>
   <div class="container-fluid py-2">
+    <div id="build-header" class="border rounded p-2 mb-3 bg-white">
```
Comment (Andrewnt219) [inline src/api/status/src/views/builds.hbs:35]: Can you make this one a [card](https://www.creative-tim.com/learning-lab/bootstrap/cards/material-dashboard)?

File: src/api/status/src/views/builds.hbs [hunk 8]
```diff
@@ -32,6 +32,43 @@
...
+                <img
+                  class="avatar avatar-xs rounded-circle"
+                  alt=""
+                  src=""
+                  id="build-sender-img"
+                />
+                <span id="build-sender-name"></span>
+              </a>
+              synchronized
+              <span><a id="build-git-sha" href="#" class="text-decoration-underline"></a></span>
```
Comment (Andrewnt219) [inline src/api/status/src/views/builds.hbs:58]: This can be `card-link`

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]

File: src/api/status/public/js/build-log/build-header.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  buildSender.setAttribute('href', githubData.sender.html_url);
+  buildSenderName.innerText = githubData.sender.login;
+  buildSenderImg.src = githubData.sender.avatar_url;
+  buildGitSHA.setAttribute('href', githubData.compare);
+  buildGitSHA.innerText = githubData.after.substring(0, 7);
+  buildResult.innerText = result === 0 ? 'Good' : 'Error';
```
Comment (Yoda-Canada) [inline src/api/status/public/js/build-log/build-header.js:19]: According to the data ,  ``` result: data.code =1```

I think the line 19 should changed to  ``` buildResult.innerText = result != 0 ? 'Good' : 'Error'; ```

File: src/api/status/src/views/builds.hbs [hunk 10]
```diff
@@ -32,6 +32,43 @@
...
+            </div>
+          </div>
+          <div
+            id="build-result"
+            class="col col-md-2 text-dark text-sm font-weight-bolder"
```
Comment (Yoda-Canada) [inline src/api/status/src/views/builds.hbs:63]: We can try to optimize the UI    
```    
     <div
            id="build-result"
            class="col col-md-2 text-dark text-sm font-weight-bolder fas fa-times-circle"
            style="color:red !important" 
          ></div>
 ```

**Comment by mqnguyen5** [comment]
I have made some adjustments based on @Kevan-Y and @Andrewnt219  reviews:

- Changed the `build-header` to use the [card layout](https://www.creative-tim.com/learning-lab/bootstrap/cards/material-dashboard), which helps adds extra padding to the left
- Used `card-link` class for the pusher's profile and the git SHA
- Removed the extra `buildHeaderTitle.innerText = 'Build Status';`

**Comment by Andrewnt219** [comment]
> I think the line 19 should changed to buildResult.innerText = result != 0 ? 'Good' : 'Error';

Yoda addressed this as well.

**Comment by mqnguyen5** [comment]
But inside the description for #2416, @humphd said:

> what the result code of the build was (e.g., 0 is "good" and non-zero is an "error")

**Review by Kevan-Y — CHANGES_REQUESTED** [review]
Could you center this?
image

**Comment by Yoda-Canada** [comment]
@humphd  I just check the issue #2416, and your comment is that "what the result code of the build was  (e.g., 0 is "good" and non-zero is an "error")" ?  According to this comment, the code ` buildResult.innerText = result === 0 ? 'Good' : 'Error'; ` is correct. 
However, my understanding is that when the build is finished, the result code of the build is good.  I checked the data of "//dev.api.telescope.cdot.systems:4000/status".  The build had been done, but the code =1. 

I am confused that although the project has been completed, the result can also be an error, right?

**Comment by humphd** [comment]
https://en.wikipedia.org/wiki/Exit_status

In Unix, a zero exit code means it worked, anything else is an error.

**Comment by Yoda-Canada** [comment]
@mqnguyen5 your code ` buildResult.innerText = result != 0 ? 'Good' : 'Error'; ` is correct. My point of view is wrong, sorry for that.

**Comment by humphd** [comment]
Good to clarify, there are lots of moving pieces here.

File: src/api/status/public/js/build-log/build-header.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+const buildSender = document.getElementById('build-sender');
+const buildSenderName = document.getElementById('build-sender-name');
+const buildSenderImg = document.getElementById('build-sender-img');
+const buildGitSHA = document.getElementById('build-git-sha');
+const buildResult = document.getElementById('build-result');
+const buildStarted = document.getElementById('build-started');
+const buildDuration = document.getElementById('build-duration');
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  buildHeaderTitle.innerHTML = '';
```
Comment (Yoda-Canada) [inline src/api/status/public/js/build-log/build-header.js:12]: Line 12 has changed, but it is different from Andrent219's comment.
image

**Review by Andrewnt219 — COMMENTED** [review]
LGTM. Can you rebase and revert the change to `autodeploymentUrl`?

**Review by Andrewnt219 — DISMISSED** [review]

File: src/api/status/public/js/build-log/build-header.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  buildSender.setAttribute('href', githubData.sender.html_url);
```
Comment (humphd) [inline src/api/status/public/js/build-log/build-header.js:13]: You don't need to use `setAttribute`, use the setter/getter on the DOM element API: `buildSender.href = ...`

File: src/api/status/public/js/build-log/build-header.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  buildSenderName.innerText = githubData.sender.login;
+  buildSenderImg.src = githubData.sender.avatar_url;
+  buildGitSHA.setAttribute('href', githubData.compare);
```
Comment (humphd) [inline src/api/status/public/js/build-log/build-header.js:16]: Same thing here

**Review by humphd — CHANGES_REQUESTED** [review]
Two small things, looks good otherwise.

Fix and rebase please.

**Review by Kevan-Y — CHANGES_REQUESTED** [review]
Not sure if it looks good this way
image
Maybe you can hide the part below on loading

You can element `hidden` of `id="build-header-info`
and have a quick check if  elem `build-header-info` is hidden or not after line 12 in `src/api/status/public/js/build-log/build-header.js` 
```js
if (buildHeaderInfo.hidden) {
    buildHeaderInfo.removeAttribute('hidden');
}
  ```

**Comment by mqnguyen5** [comment]
Changes have been made in response to @humphd and @Kevan-Y reviews. Sorry for not able to complete this PR sooner, I should've schedule my time better.

**Review by Andrewnt219 — APPROVED** [review]
Looks great. Well done 👍.

**Review by Yoda-Canada — APPROVED** [review]
Great！You've really done hard work.

**Review by Kevan-Y — APPROVED** [review]
LGTM!

**Pull request merged**